lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -999,9 +1005,27 @@ newtype SlotNo = SlotNo { unSlotNo :: Word32 }
     deriving stock (Show, Read, Eq, Ord, Generic)
     deriving newtype (Num, Buildable, NFData, Enum)
 
-newtype EpochNo = EpochNo { unEpochNo :: Word64 }
+newtype EpochNo = EpochNo { unEpochNo :: Word31 }
     deriving stock (Show, Read, Eq, Ord, Generic)
-    deriving newtype (Num, Buildable, NFData, Enum)
+    deriving newtype (Num, Bounded, Enum)
+
+instance Buildable EpochNo where
+    build (EpochNo e) = build $ fromIntegral @Word31 @Word32 e
+
+instance NFData EpochNo where
+    rnf (EpochNo !_) = ()
+
+-- | Convert the specified value into an 'EpochNo', or fail if the value is
+--   too large.
+unsafeEpochNo :: HasCallStack => Word32 -> EpochNo
+unsafeEpochNo epochNo
+    | epochNo > maxEpochNo =
+        error "unsafeEpochNo: epoch number out of bounds."
```

Comment:
**[suggestion: user-experience]**
The error message could include the actual bounds and what was the given value.

**Why**
Increase a bit the user experience and debugging capability.